### PR TITLE
Use std::chrono::duration in bigtable::ClientOptions::SetGrpclbFallbackTimeout

### PR DESCRIPTION
**Pull request by hemant-qlogic**
Changed Test case to use the duration instead of int in method SetGrpclbFallbackTimeout

### CODE REVIEW

File: bigtable/client/client_options.h [hunk 1]
```diff
@@ -92,6 +92,27 @@ class ClientOptions {
...
   }
 
+  /**
+   * Set the grpclb fallback timeout (std::chrono::duration<> duration) for the
+   * channel.
+   *
+   * Please see the docs for grpc::ChannelArguments::SetGrpclbFallbackTimeout()
+   * on https://grpc.io/grpc/cpp/classgrpc_1_1_channel_arguments.html
+   * for more details.
+   *
```
Comment (coryan) [inline bigtable/client/client_options.h]: Please document the `Rep` and `Period` template parameters.  Look here:

https://github.com/GoogleCloudPlatform/google-cloud-cpp/blob/master/bigtable/client/filters.h#L168

for an example.

File: bigtable/client/client_options.h [hunk 2]
```diff
@@ -92,6 +92,27 @@ class ClientOptions {
...
+   */
+  template <typename Rep, typename Period>
+  void SetGrpclbFallbackTimeout(
+      std::chrono::duration<Rep, Period> fallback_timeout) {
+    int fallback_timeout_ms = 0;
```
Comment (coryan) [inline bigtable/client/client_options.h]: Please move the variable definition to the point where it is initialized (line 111).

File: bigtable/client/client_options.h [hunk 3]
```diff
@@ -92,6 +92,27 @@ class ClientOptions {
...
+
+    std::chrono::milliseconds ft_ms =
+        std::chrono::duration_cast<std::chrono::milliseconds>(fallback_timeout);
+    fallback_timeout_ms = static_cast<std::int32_t>(ft_ms.count());
```
Comment (coryan) [inline bigtable/client/client_options.h]: What if `ft_ms.count()` is larger than what fits in a `std::int32_t`?  The documentation should say what happens and we should raise an exception (probably `std::out_of_range`).

File: bigtable/client/client_options.h [hunk 4]
```diff
@@ -92,6 +92,27 @@ class ClientOptions {
     channel_arguments_.SetGrpclbFallbackTimeout(fallback_timeout);
   }
 
+  /**
```
Comment (coryan) [inline bigtable/client/client_options.h]: Please remove the previous `SetGrpcFallbackTimeout(int)` overload.

File: bigtable/client/client_options_test.cc [hunk 5]
```diff
@@ -112,7 +112,7 @@ TEST(ClientOptionsTest, EditCredentials) {
 
 TEST(ClientOptionsTest, SetGrpclbFallbackTimeout) {
   bigtable::ClientOptions client_options_object = bigtable::ClientOptions();
-  client_options_object.SetGrpclbFallbackTimeout(5);
+  client_options_object.SetGrpclbFallbackTimeout(std::chrono::milliseconds(5));
```
Comment (coryan) [inline bigtable/client/client_options_test.cc]: Please add a new unit test that uses `std::chrono::seconds()` too.

**Comment by coryan** [comment]
In general, if you refer to the issue in the body of your PR description (e.g., #169) github will link the two.  No need to link it from the PR title.

File: bigtable/client/client_options_test.cc (hunk 5 above)
Comment (hemant-qlogic) [reply bigtable/client/client_options_test.cc]: Shall I add unit test for std::out_of_range exception also?

Comment (coryan) [reply bigtable/client/client_options_test.cc]: Good point, yes, please do.

File: bigtable/client/client_options.h [hunk 6]
```diff
@@ -81,15 +82,56 @@ class ClientOptions {
...
-   * Set the grpclb fallback timeout (in ms) for the channel.
+   * Set the grpclb fallback timeout with the timestamp [@p fallback_timeout)
+   * for the channel.
+   *
+   * This function accepts any instantiation of 'std::chrono::duration<>' for
+   * @p duration parameter convert it into milliseconds and pass it to
+   * channel_arguments. For example:
+   *
+   * @code
+   * using namespace std::chrono_literals; // C++14
```
Comment (coryan) [inline bigtable/client/client_options.h]: Why introduced the C++14 namespace if you do not use it?  Please change the code to use the chrono literals, or simply remove this line.

File: bigtable/client/client_options.h [hunk 7]
```diff
@@ -81,15 +82,56 @@ class ClientOptions {
...
+   * of int32. The code will throw exception std::out_of_range if range goes
+   * outside of int32.
+   *
+   * @tparam Rep a placeholder to match the Rep tparam for @p fallback_timeout,
+   *     the semantics of this template parameter are documented in
+   *     `std::chrono::duration<>` (in brief, the underlying arithmetic type
+   *     used to store the number of ticks), for our purposes it is simply a
+   *     formal parameter.
+   * @tparam Period a placeholder to match the Period tparam for @p
+   * fallback_timeout
```
Comment (coryan) [inline bigtable/client/client_options.h]: Please indent the paragraph correctly.  If `fallback_timeout` does not fit in the previous line it should be indented by 4 spaces and the rest of the paragraph needs to be reflowed.

File: bigtable/client/client_options.h [hunk 8]
```diff
@@ -81,15 +82,56 @@ class ClientOptions {
...
+  template <typename Rep, typename Period>
+  void SetGrpclbFallbackTimeout(
+      std::chrono::duration<Rep, Period> fallback_timeout) {
+    std::chrono::milliseconds ft_ms =
+        std::chrono::duration_cast<std::chrono::milliseconds>(fallback_timeout);
+
+    if (ft_ms.count() > std::numeric_limits<int32_t>::max())
+      absl::base_internal::ThrowStdOutOfRange("Duration Exceeds Range for int");
+
+    int32_t fallback_timeout_ms = static_cast<std::int32_t>(ft_ms.count());
```
Comment (coryan) [inline bigtable/client/client_options.h]: Use `auto fallback_timeout_ms`, the type is obvious from the `static_cast`.

File: bigtable/client/client_options.h [hunk 9]
```diff
@@ -81,15 +82,56 @@ class ClientOptions {
...
-  void SetGrpclbFallbackTimeout(int fallback_timeout) {
-    channel_arguments_.SetGrpclbFallbackTimeout(fallback_timeout);
+  template <typename Rep, typename Period>
+  void SetGrpclbFallbackTimeout(
+      std::chrono::duration<Rep, Period> fallback_timeout) {
+    std::chrono::milliseconds ft_ms =
+        std::chrono::duration_cast<std::chrono::milliseconds>(fallback_timeout);
+
+    if (ft_ms.count() > std::numeric_limits<int32_t>::max())
+      absl::base_internal::ThrowStdOutOfRange("Duration Exceeds Range for int");
```
Comment (coryan) [inline bigtable/client/client_options.h]: We should not be using `absl::*internal`.  Please use a plain `throw std::out_of_range` here.  We will fix all of those in a future PR and introduce our own `bigtable::internal::Throw*` functions (tedious, but needs to be done).

File: bigtable/client/client_options_test.cc [hunk 10]
```diff
@@ -122,6 +123,26 @@ TEST(ClientOptionsTest, SetGrpclbFallbackTimeout) {
   // 2nd element of test_args.
   EXPECT_EQ(GRPC_ARG_GRPCLB_FALLBACK_TIMEOUT_MS,
             grpc::string(test_args.args[1].key));
+
+  // Test seconds are converted into milliseconds
```
Comment (coryan) [inline bigtable/client/client_options_test.cc]: Please separate these assertions to their own `TEST(ClientOptionsTest, ...)` functions.

File: bigtable/client/client_options.h [hunk 11]
```diff
@@ -81,15 +82,56 @@ class ClientOptions {
   }
 
   /**
-   * Set the grpclb fallback timeout (in ms) for the channel.
+   * Set the grpclb fallback timeout with the timestamp [@p fallback_timeout)
```
Comment (coryan) [inline bigtable/client/client_options.h]: Please remove the open bracket `[` and close paren `)`.

File: bigtable/client/client_options.h [hunk 12]
```diff
@@ -15,6 +15,7 @@
 #ifndef GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_CLIENT_OPTIONS_H_
 #define GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_CLIENT_OPTIONS_H_
 
+#include "absl/base/internal/throw_delegate.h"
```
Comment (coryan) [inline bigtable/client/client_options.h]: We should never directly include headers from `absl/*/internal/*`, please remove.

File: bigtable/client/client_options.h [hunk 13]
```diff
@@ -81,15 +82,56 @@ class ClientOptions {
...
    */
-  void SetGrpclbFallbackTimeout(int fallback_timeout) {
-    channel_arguments_.SetGrpclbFallbackTimeout(fallback_timeout);
+  template <typename Rep, typename Period>
+  void SetGrpclbFallbackTimeout(
+      std::chrono::duration<Rep, Period> fallback_timeout) {
+    std::chrono::milliseconds ft_ms =
+        std::chrono::duration_cast<std::chrono::milliseconds>(fallback_timeout);
+
+    if (ft_ms.count() > std::numeric_limits<int32_t>::max())
```
Comment (coryan) [inline bigtable/client/client_options.h]: Please use `std::int32_t`, we do not want to rely on some header magically inserting the `int*_t` types into the global namespace.

File: bigtable/client/client_options.h (hunk 9 above)
Comment (coryan) [inline bigtable/client/client_options.h]: s/Duration Exceeds Range for int/Duration exceeds range for int/

File: bigtable/client/client_options.h (hunk 13 above)
Comment (coryan) [reply bigtable/client/client_options.h]: Actually, `grpc::ChannelArguments::SetGrpclbFallbackTimeout` takes an `int` parameter, so we should not be using `std::int32_t` here.

File: bigtable/client/client_options.h (hunk 8 above)
Comment (coryan) [reply bigtable/client/client_options.h]: And we should be doing `static_cast<int>`

File: bigtable/client/client_options.h (hunk 7 above)
Comment (coryan) [reply bigtable/client/client_options.h]: This is still pending, the rest looks good, please fix so we can merge your changes.

**Pull request merged**